### PR TITLE
Add distinctBy operation

### PR DESCRIPTION
**Pull request by marcelocenerine**
Addresses #239:
- Add `distinctBy` operation and re-write `distinct` in terms of `distinctBy`
- Add unit tests and benchmarks for `distinctBy`

### CODE REVIEW

File: src/main/scala/strawman/collection/Iterator.scala [hunk 1]
```diff
@@ -469,13 +469,25 @@ trait Iterator[+A] extends IterableOnce[A] { self =>
...
-    private val traversedValues = mutable.HashSet.empty[A]
+  /**
+    *  Builds a new iterator from this one without any duplicated elements as determined by `==` after applying
+    *  the transforming function `f`.
+    *
+    *  @param f The transforming function whose result is used to determine the uniqueness of each element
+    *  @tparam B the type of the elements after being transformed by `f`
+    *  @return iterator with distinct elements
+    *
+    *  @note   Reuse: $consumesIterator
```
Comment (marcelocenerine) [inline src/main/scala/strawman/collection/Iterator.scala]: does this @note have any relevance?

**Comment by jozic** [comment]
finally this is happening :)

**Comment by marcelocenerine** [comment]
* current impl. of `distinct`:

transform_distinct



* `distinct` implemented in terms of `distinctBy`:

transform_distinctold

**Comment by julienrf** [comment]
@marcelocenerine Do you also have the actual numbers of the benchmarks? It is hard to read them on the images. Alternatively, you could generate one image per collection type (`List`, `Vector`, etc.) in which you would have two lines, one for the current implementation and one for the implementation of this PR.

Also, the title of the second image (“transform_distinctOld”) makes me think that it shows the performance of the current implementation (not the one of this PR), can you double-check?

**Comment by marcelocenerine** [comment]
@julienrf, as you requested I attached a [document](https://github.com/scala/collection-strawman/files/1326568/distinct.benchmarks.pdf) showing the actual numbers as well as individual charts for each collection type. The classes with the "Current" prefix refers to the current implementation of `distinct` whereas the ones without it refers to the implementation of `distinct` in terms of `distinctBy`. The numbers are pretty much identical for all collections types.

**Comment by julienrf** [comment]
Thanks a lot @marcelocenerine!

**Pull request merged**